File: app/src/main/java/com/bodyweight/fitness/model/Routine.java
package com.bodyweight.fitness.model;

import com.bodyweight.fitness.model.json.JSONRoutine;

import java.io.Serializable;
import java.util.ArrayList;

import io.mazur.glacier.Glacier;

public class Routine implements Serializable {
    private ArrayList<Category> mCategories = new ArrayList<>();
    private ArrayList<Section> mSections = new ArrayList<>();
    private ArrayList<Exercise> mExercises = new ArrayList<>();
    private ArrayList<Exercise> mLinkedExercises = new ArrayList<>();
    private ArrayList<LinkedRoutine> mLinkedRoutine = new ArrayList<>();

    public Routine(JSONRoutine JSONRoutine) {
        Category currentCategory = null;
        Section currentSection = null;
        Exercise currentExercise = null;

        for(com.bodyweight.fitness.model.json.JSONLinkedRoutine JSONLinkedRoutine : JSONRoutine.getPartRoutines()) {
            if(JSONLinkedRoutine.getType() == RoutineType.CATEGORY) {
                currentCategory = new Category(
                        JSONLinkedRoutine.getCategoryId(),
                        JSONLinkedRoutine.getTitle()
                );

                mCategories.add(currentCategory);

                /**
                 * Whether to show categories in the linked routine (e.g. Drawer).
                 */
                mLinkedRoutine.add(currentCategory);
            }

            /**
             * Updated sections for each category and add into linked routine.
             */
            else if(JSONLinkedRoutine.getType() == RoutineType.SECTION) {
                currentSection = new Section(
                        JSONLinkedRoutine.getSectionId(),
                        JSONLinkedRoutine.getTitle(),
                        JSONLinkedRoutine.getDescription(),
                        JSONLinkedRoutine.getMode()
                );

                currentCategory.insertSection(currentSection);

                mSections.add(currentSection);
                mLinkedRoutine.add(currentSection);
            }

            /**
             * Update exercises for each section, link them together.
             */
            else if(JSONLinkedRoutine.getType() == RoutineType.EXERCISE) {
                Exercise exercise = new Exercise(
                        JSONLinkedRoutine.getExerciseId(),
                        JSONLinkedRoutine.getId(),
                        JSONLinkedRoutine.getLevel(),
                        JSONLinkedRoutine.getTitle(),
                        JSONLinkedRoutine.getDescription(),
                        JSONLinkedRoutine.getYouTubeId(),
                        JSONLinkedRoutine.getDefaultSet()
                );

                /**
                 * Add exercise into a section.
                 */
                currentSection.insertExercise(exercise);

                /**
                 * List contains all exercises, no matter what level.
                 */
                mExercises.add(exercise);

                /**
                 * Add only first level of exercise when linking.
                 */
                if(currentSection.getSectionMode() == SectionMode.LEVELS || currentSection.getSectionMode() == SectionMode.PICK) {
                    String currentExerciseId = Glacier.get(currentSection.getSectionId(), String.class);

                    if(currentExerciseId != null) {
                        if(exercise.getExerciseId().matches(currentExerciseId)) {
                            mLinkedExercises.add(exercise);
                            mLinkedRoutine.add(exercise);

                            exercise.setPrevious(currentExercise);

                            if(currentExercise != null) {
                                currentExercise.setNext(exercise);
                            }

                            currentExercise = exercise;
                            currentSection.setCurrentLevel(exercise);
                        }
                    } else {
                        if(currentSection.getExercises().size() == 1) {
                            mLinkedExercises.add(exercise);
                            mLinkedRoutine.add(exercise);

                            exercise.setPrevious(currentExercise);

                            if(currentExercise != null) {
                                currentExercise.setNext(exercise);
                            }

                            currentExercise = exercise;
                        }
                    }
                } else {
                    exercise.setPrevious(currentExercise);

                    if(currentExercise != null) {
                        currentExercise.setNext(exercise);
                    }

                    currentExercise = exercise;

                    mLinkedExercises.add(exercise);
                    mLinkedRoutine.add(exercise);
                }
            }
        }
    }

    public ArrayList<Category> getCategories() {
        return mCategories;
    }

    public ArrayList<Section> getSections() {
        return mSections;
    }

    public ArrayList<Exercise> getExercises() {
        return mExercises;
    }

    public ArrayList<Exercise> getLinkedExercises() {
        return mLinkedExercises;
    }

    public void setLevel(Exercise exercise, int level) {
        Exercise currentSectionExercise = exercise.getSection().getCurrentExercise();

        if(currentSectionExercise != exercise) {
            if(currentSectionExercise.getPrevious() != null) {
                currentSectionExercise.getPrevious().setNext(exercise);
            }

            if(currentSectionExercise.getNext() != null) {
                currentSectionExercise.getNext().setPrevious(exercise);
            }

            exercise.setPrevious(currentSectionExercise.getPrevious());
            exercise.setNext(currentSectionExercise.getNext());

            exercise.getSection().setCurrentLevel(level);

            currentSectionExercise.setPrevious(null);
            currentSectionExercise.setNext(null);

            int indexOfCurrentExercise = mLinkedRoutine.indexOf(currentSectionExercise);
            if(indexOfCurrentExercise == -1) {
                return;
            }

            mLinkedRoutine.set(indexOfCurrentExercise, exercise);
        }
    }

    public ArrayList<LinkedRoutine> getLinkedRoutine() {
        return mLinkedRoutine;
    }
}

File: app/src/main/kotlin/com/bodyweight/fitness/model/Routine.kt
package com.bodyweight.fitness.model

import com.bodyweight.fitness.persistence.Glacier
import java.io.Serializable
import java.util.*

abstract class LinkedRoutine : Serializable {
    abstract val title: String
    abstract val type: RoutineType
}

class Routine(JSONRoutine: JSONRoutine) : Serializable {
    var routineId: String = "routine0"
    var title: String = ""
    var subtitle: String = ""

    val categories = ArrayList<Category>()
    val sections = ArrayList<Section>()
    val exercises = ArrayList<Exercise>()
    val linkedExercises = ArrayList<Exercise>()
    val linkedRoutine = ArrayList<LinkedRoutine>()

    init {
        routineId = JSONRoutine.routineId
        title = JSONRoutine.title
        subtitle = JSONRoutine.subtitle

        var currentCategory: Category? = null
        var currentSection: Section? = null
        var currentExercise: Exercise? = null

        for (JSONLinkedRoutine in JSONRoutine.routine) {
            if (JSONLinkedRoutine.routineType === RoutineType.Category) {
                currentCategory = Category(JSONLinkedRoutine.categoryId, JSONLinkedRoutine.title)

                categories.add(currentCategory)
                linkedRoutine.add(currentCategory)
            } else if (JSONLinkedRoutine.routineType === RoutineType.Section) {
                currentSection = Section(
                        JSONLinkedRoutine.sectionId,
                        JSONLinkedRoutine.title,
                        JSONLinkedRoutine.description,
                        JSONLinkedRoutine.sectionMode)

                currentCategory!!.insertSection(currentSection)

                sections.add(currentSection)
                linkedRoutine.add(currentSection)
            } else if (JSONLinkedRoutine.routineType === RoutineType.Exercise) {
                val exercise = Exercise(
                        JSONLinkedRoutine.exerciseId,
                        JSONLinkedRoutine.level,
                        JSONLinkedRoutine.title,
                        JSONLinkedRoutine.description,
                        JSONLinkedRoutine.youTubeId,
                        JSONLinkedRoutine.videoId,
                        JSONLinkedRoutine.defaultSet)

                currentSection!!.insertExercise(exercise)

                exercises.add(exercise)

                if (currentSection.sectionMode === SectionMode.Levels || currentSection.sectionMode === SectionMode.Pick) {
                    val currentExerciseId = Glacier.get(currentSection.sectionId, String::class.java)

                    if (currentExerciseId != null) {
                        if (exercise.exerciseId.matches(currentExerciseId.toRegex())) {
                            linkedExercises.add(exercise)
                            linkedRoutine.add(exercise)

                            exercise.previous = currentExercise

                            if (currentExercise != null) {
                                currentExercise.next = exercise
                            }

                            currentExercise = exercise
                            currentSection.setCurrentLevel(exercise)
                        }
                    } else {
                        if (currentSection.exercises.size == 1) {
                            linkedExercises.add(exercise)
                            linkedRoutine.add(exercise)

                            exercise.previous = currentExercise

                            if (currentExercise != null) {
                                currentExercise.next = exercise
                            }

                            currentExercise = exercise
                        }
                    }
                } else {
                    exercise.previous = currentExercise

                    if (currentExercise != null) {
                        currentExercise.next = exercise
                    }

                    currentExercise = exercise

                    linkedExercises.add(exercise)
                    linkedRoutine.add(exercise)
                }
            }
        }
    }

    fun setLevel(exercise: Exercise, level: Int) {
        val currentSectionExercise = exercise.section!!.currentExercise

        if (currentSectionExercise != exercise) {
            if (currentSectionExercise.previous != null) {
                currentSectionExercise.previous!!.next = exercise
            }

            if (currentSectionExercise.next != null) {
                currentSectionExercise.next!!.previous = exercise
            }

            exercise.previous = currentSectionExercise.previous
            exercise.next = currentSectionExercise.next

            exercise.section!!.currentLevel = level

            currentSectionExercise.previous = null
            currentSectionExercise.next = null

            val indexOfCurrentExercise = linkedRoutine.indexOf(currentSectionExercise)
            if (indexOfCurrentExercise > -1) {
                linkedRoutine[indexOfCurrentExercise] = exercise
            }

            val indexOfLinkedExercise = linkedExercises.indexOf(currentSectionExercise)
            if (indexOfLinkedExercise > -1) {
                linkedExercises[indexOfLinkedExercise] = exercise
            }
        }
    }
}

class Category(
        val categoryId: String,
        val categoryTitle: String
) : LinkedRoutine(), Serializable {
    val sections = ArrayList<Section>()

    override val title: String = categoryTitle
    override val type: RoutineType = RoutineType.Category

    fun insertSection(section: Section) {
        section.category = this

        sections.add(section)
    }
}

class Section(
        val sectionId: String,
        val sectionTitle: String,
        val description: String,
        val sectionMode: SectionMode) : LinkedRoutine(), Serializable {

    var category: Category? = null
    var currentLevel = 0
        set(level) {
            if (sectionMode == SectionMode.Levels || sectionMode == SectionMode.Pick) {
                if (level < 0 || level >= exercises.size) {
                    return
                }

                field = level
            }
        }

    val exercises = ArrayList<Exercise>()

    override val title: String
        get() = sectionTitle

    override val type: RoutineType
        get() = RoutineType.Section

    fun setCurrentLevel(exercise: Exercise) {
        var current = 0
        for (ex in exercises) {
            if (ex == exercise) {
                currentLevel = current
                return
            }

            current++
        }
    }

    val availableLevels: Int
        get() {
            if (sectionMode == SectionMode.Levels || sectionMode == SectionMode.Pick) {
                return exercises.size
            } else {
                return 0
            }
        }

    fun insertExercise(exercise: Exercise) {
        exercise.category = category
        exercise.section = this

        exercises.add(exercise)
    }

    val currentExercise: Exercise
        get() = exercises[currentLevel]
}

class Exercise(
        val exerciseId: String,
        val level: String,
        val exerciseTitle: String,
        val description: String,
        val youTubeId: String,
        val videoId: String,
        val defaultSet: String) : LinkedRoutine(), Serializable {

    var category: Category? = null
    var section: Section? = null

    var previous: Exercise? = null
    var next: Exercise? = null

    override val title: String
        get() = exerciseTitle

    override val type: RoutineType
        get() = RoutineType.Exercise

    val isTimedSet: Boolean
        get() = defaultSet.equals("timed", ignoreCase = true)

    val hasProgressions: Boolean
        get() = section?.sectionMode != SectionMode.All

    val isPrevious: Boolean
        get() = previous != null

    val isNext: Boolean
        get() = next != null
}